File: Common/Changes.cpp
#include "Changes.h"
#include "Connection.h"
#include <algorithm>

using namespace std;

const char * kDefaultListRevision = "-1";
const char * kNewListRevision = "-2";

Changelist::Changelist() { }
// Changelist(std::string const& description);

ChangelistRevision Changelist::GetRevision() const 
{
	return m_Revision;
}

void Changelist::SetRevision(const ChangelistRevision& revison) 
{
	m_Revision = revison;
}	

std::string Changelist::GetDescription() const 
{
	return m_Description;
}

void Changelist::SetDescription(const std::string& description) 
{
	m_Description = description;
} 

std::string Changelist::GetTimestamp() const
{
	return m_Timestamp;
}

void Changelist::SetTimestamp(const std::string& timestamp)
{
	m_Timestamp = timestamp;
}

std::string Changelist::GetCommitter() const
{
	return m_Committer;
}

void Changelist::SetCommitter(const std::string& committer)
{
	m_Committer = committer;
}

//Connection& operator<<(Connection& p, ChangelistRevision revision)
//{
//	p << revision.c_str();
//	return p;
//}

Connection& operator>>(Connection& p, ChangelistRevision& revision)
{
	string line;
	p.ReadLine(line);
	revision = line;
	return p;
}

Connection& operator<<(Connection& p, const Changelist& changelist)
{
	p.DataLine(changelist.GetRevision().c_str());
	p.DataLine(changelist.GetDescription());
	return p;
}

Connection& operator>>(Connection& p, Changelist& cl)
{
	string line;
	p.ReadLine(line);
	cl.SetRevision(line);
	p.ReadLine(line); 
	cl.SetDescription(line);
	return p;
}

File: TFSPlugin/TfsPlugin/Common/Changes.cs
using System.Collections.Generic;

namespace TfsPlugin
{
    public class ChangelistRevision : IReadsFromConnection
    {
        public readonly static string kDefaultListRevision = "-1";
        public readonly static string kNewListRevision = "-2";

        public string Value { get; set; }

        public ChangelistRevision()
        {

        }

        public ChangelistRevision(string value)
        {
            this.Value = value;
        }

        public static implicit operator string(ChangelistRevision p)
        {
            return p.Value;
        }

        public static implicit operator ChangelistRevision(string path)
        {
            return new ChangelistRevision(path);
        }

        public void Read(Connection p)
        {
            Value = p.ReadLine();
        }

        public int ToInt()
        {
            return int.Parse(Value);
        }
    }

    public class Changelist : IReadsFromConnection, IWritesToConnection
    {
        public Changelist()
        {
            Revision = string.Empty;
            Description = string.Empty;
            Timestamp = string.Empty;
            Committer = string.Empty;
        }

        public ChangelistRevision GetRevision() { return Revision; }
        public void SetRevision(ChangelistRevision revison) { Revision = revison; }

        public string GetDescription() { return Description; }
        public void SetDescription(string description) { Description = description; }

        public string GetTimestamp() { return Timestamp; }
        public void SetTimestamp(string timestamp) { Timestamp = timestamp; }

        public string GetCommitter() { return Committer; }
        public void SetCommitter(string committer) { Committer = committer; }

        public ChangelistRevision Revision { get; set; }
        public string Description { get; set; }
        public string Timestamp { get; set; }
        public string Committer { get; set; }

        public void Write(Connection p)
        {
            p.DataLine(GetRevision());
            p.DataLine(GetDescription());
        }

        public void Read(Connection p)
        {
            SetRevision(p.ReadLine());
            SetDescription(p.ReadLine());
        }
    }

    public class ChangelistRevisions : List<ChangelistRevision>
    {

    }
}
